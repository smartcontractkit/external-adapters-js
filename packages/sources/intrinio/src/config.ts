import { Requester } from '@chainlink/ea-bootstrap'
import { Config } from '@chainlink/types'

export const NAME = 'INTRINIO'

export const DEFAULT_API_ENDPOINT = 'https://api-v2.intrinio.com/'
// const DEFAULT_WS_API_ENDPOINT = ''

export const makeConfig = (prefix?: string): Config => {
  const config = Requester.getDefaultConfig(prefix, true)
  config.api.baseURL = config.api.baseURL || DEFAULT_API_ENDPOINT
  return config
}
