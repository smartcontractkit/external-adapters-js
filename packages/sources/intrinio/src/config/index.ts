import { Requester } from '@chainlink/ea-bootstrap'
import { Config } from '@chainlink/ea-bootstrap'

export const NAME = 'INTRINIO'

export const DEFAULT_BASE_URL = 'https://api-v2.intrinio.com/'
// const DEFAULT_WS_API_ENDPOINT = ''

export const makeConfig = (prefix?: string): Config => {
  const config = Requester.getDefaultConfig(prefix, true)
  config.api.baseURL = config.api.baseURL || DEFAULT_BASE_URL
  return config
}
