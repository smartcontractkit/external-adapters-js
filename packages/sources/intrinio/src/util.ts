/**
 * This code is directly copied from Intrinio's library.  We have a ticket to refactor this file.
 */

import * as https from 'https'
import * as events from 'events'
import { Logger } from '@chainlink/ea-bootstrap'

const EventEmitter = events.EventEmitter

const SELF_HEAL_BACKOFFS = [0, 100, 500, 1000, 2000]

export class IntrinioRealtime extends EventEmitter {
  options: any
  token: null | string = null
  websocket = null
  ready = false
  channels = {}
  joinedChannels = {}
  afterConnected = null // Promise
  self_heal_backoff = Array.from(SELF_HEAL_BACKOFFS)
  self_heal_ref = null
  quote_callback = null
  error_callback = null

  constructor(options: any) {
    super()

    this.options = options
    this.token = null
    this.websocket = null
    this.ready = false
    this.channels = {}
    this.joinedChannels = {}
    this.afterConnected = null // Promise
    this.self_heal_backoff = Array.from(SELF_HEAL_BACKOFFS)
    this.self_heal_ref = null
    this.quote_callback = null
    this.error_callback = null

    // Parse options
    if (!options) {
      this._throw('Need a valid options parameter')
    }

    if (options.api_key) {
      if (!this._validAPIKey(options.api_key)) {
        this._throw('API Key was formatted invalidly')
      }
    } else {
      if (!options.username && !options.password) {
        this._throw('API key or username and password are required')
      }

      if (!options.username) {
        this._throw('Need a valid username')
      }

      if (!options.password) {
        this._throw('Need a valid password')
      }
    }

    const providers = ['iex', 'quodd', 'cryptoquote', 'fxcm']
    if (!options.provider || !providers.includes(options.provider)) {
      this._throw('Need a valid provider: iex, quodd, cryptoquote, or fxcm')
    }
  }

  _throw(e: any) {
    let handled = false
    if (typeof e === 'string') {
      e = 'IntrinioRealtime | ' + e
    }
    if (this.listenerCount('error') > 0) {
      Logger.error(e)
      handled = true
    }
    if (!handled) {
      throw e
    }
  }

  _makeAuthUrl(): { host: string; path: string } {
    let auth_url = {
      host: '',
      path: '',
    }

    if (this.options.provider == 'iex') {
      auth_url = {
        host: 'realtime.intrinio.com',
        path: '/auth',
      }
    } else if (this.options.provider == 'quodd') {
      auth_url = {
        host: 'api.intrinio.com',
        path: '/token?type=QUODD',
      }
    } else if (this.options.provider == 'cryptoquote') {
      auth_url = {
        host: 'crypto.intrinio.com',
        path: '/auth',
      }
    } else if (this.options.provider == 'fxcm') {
      auth_url = {
        host: 'fxcm.intrinio.com',
        path: '/auth',
      }
    }

    if (this.options.api_key) {
      auth_url = this._makeAPIAuthUrl(auth_url)
    }

    return auth_url
  }

  _makeAPIAuthUrl(auth_url: any): any {
    let path = auth_url.path

    if (path.includes('?')) {
      path = path + '&'
    } else {
      path = path + '?'
    }

    auth_url.path = path + 'api_key=' + this.options.api_key
    return auth_url
  }

  _makeHeaders() {
    if (this.options.api_key) {
      return {
        'Content-Type': 'application/json',
      }
    } else {
      const { username, password } = this.options

      return {
        'Content-Type': 'application/json',
        Authorization: 'Basic ' + Buffer.from(username + ':' + password).toString('base64'),
      }
    }
  }

  _refreshToken() {
    Logger.debug('Requesting auth token...')

    return new Promise<void>((fulfill, reject) => {
      const agent = this.options.agent || false
      const { host, path } = this._makeAuthUrl()
      const headers = this._makeHeaders()

      // Get token
      const options = {
        host: host,
        path: path,
        agent: agent,
        headers: headers,
      }

      const req = https.get(options, (res: any) => {
        if (res.statusCode == 401) {
          this._throw('Unable to authorize')
          reject()
        } else if (res.statusCode != 200) {
          console.error(
            'IntrinioRealtime | Could not get auth token: Status code ' + res.statusCode,
          )
          reject()
        } else {
          res.on('data', (data: any) => {
            this.token = Buffer.from(data, 'base64').toString()
            Logger.debug('Received auth token!')
            fulfill()
          })
        }
      })

      req.on('error', (e) => {
        console.error('IntrinioRealtime | Could not get auth token: ' + e)
        reject(e)
      })

      req.end()
    })
  }

  async _makeSocketUrl(): Promise<string> {
    if (!this.token) {
      await this._refreshToken()
      if (!this.token) return ''
    }

    if (this.options.provider == 'iex') {
      return (
        'wss://realtime.intrinio.com/socket/websocket?vsn=1.0.0&token=' +
        encodeURIComponent(this.token)
      )
    } else if (this.options.provider == 'quodd') {
      return 'wss://www5.quodd.com/websocket/webStreamer/intrinio/' + encodeURIComponent(this.token)
    } else if (this.options.provider == 'cryptoquote') {
      return (
        'wss://crypto.intrinio.com/socket/websocket?vsn=1.0.0&token=' +
        encodeURIComponent(this.token)
      )
    } else if (this.options.provider == 'fxcm') {
      return (
        'wss://fxcm.intrinio.com/socket/websocket?vsn=1.0.0&token=' + encodeURIComponent(this.token)
      )
    }

    return ''
  }

  _makeHeartbeatMessage(): any {
    if (this.options.provider == 'quodd') {
      return { event: 'heartbeat', data: { action: 'heartbeat', ticker: Date.now() } }
    } else if (['iex', 'cryptoquote', 'fxcm'].includes(this.options.provider)) {
      return { topic: 'phoenix', event: 'heartbeat', payload: {}, ref: null }
    }
  }

  _makeJoinMessage(channel: string): any {
    if (this.options.provider == 'iex') {
      return {
        topic: this._parseIexTopic(channel),
        event: 'phx_join',
        payload: {},
        ref: null,
      }
    } else if (this.options.provider == 'quodd') {
      return {
        event: 'subscribe',
        data: {
          ticker: channel,
          action: 'subscribe',
        },
      }
    } else if (['cryptoquote', 'fxcm'].includes(this.options.provider)) {
      return {
        topic: channel,
        event: 'phx_join',
        payload: {},
        ref: null,
      }
    }
  }

  _makeLeaveMessage(channel: string): any {
    if (this.options.provider == 'iex') {
      return {
        topic: this._parseIexTopic(channel),
        event: 'phx_leave',
        payload: {},
        ref: null,
      }
    } else if (this.options.provider == 'quodd') {
      return {
        event: 'unsubscribe',
        data: {
          ticker: channel,
          action: 'unsubscribe',
        },
      }
    } else if (['cryptoquote', 'fxcm'].includes(this.options.provider)) {
      return {
        topic: channel,
        event: 'phx_leave',
        payload: {},
        ref: null,
      }
    }
  }

  _parseIexTopic(channel: string): string {
    let topic = null
    if (channel == '$lobby') {
      topic = 'iex:lobby'
    } else if (channel == '$lobby_last_price') {
      topic = 'iex:lobby:last_price'
    } else {
      topic = 'iex:securities:' + channel
    }
    return topic
  }

  _validAPIKey(api_key: string): boolean {
    if (typeof api_key !== 'string') {
      return false
    }

    if (api_key === '') {
      return false
    }

    return true
  }
}
