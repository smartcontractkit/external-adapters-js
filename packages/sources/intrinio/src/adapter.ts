import { Requester, Validator } from '@chainlink/ea-bootstrap'
import { AdapterRequest, AdapterResponse, Config, MakeWSHandler } from '@chainlink/types'
import IntrinioRealtime from 'intrinio-realtime'
import { makeConfig, NAME } from './config'

export const customParams = {
  base: ['base', 'from', 'asset'],
}

export const execute = async (input: AdapterRequest, config: Config) => {
  const validator = new Validator(input, customParams)
  if (validator.error) throw validator.error

  const jobRunID = validator.validated.id
  const symbol = validator.validated.data.base.toUpperCase()

  const url = `securities/${symbol}/prices/realtime`
  const params = {
    api_key: config.apiKey,
  }

  const request = {
    ...config.api,
    url,
    params,
  }

  const response = await Requester.request(request)
  response.data.result = Requester.validateResultNumber(response.data, ['last_price'])

  return Requester.success(jobRunID, response)
}

export const makeExecute = (config?: Config) => {
  return async (request: AdapterRequest) => execute(request, config || makeConfig())
}

export const makeWSHandler = (config?: Config): MakeWSHandler => {
  // https://github.com/intrinio/intrinio-realtime-node-sdk

  const getBase = (input: AdapterRequest): string => {
    const validator = new Validator(input, customParams, {}, false)
    if (validator.error) {
      return ''
    }
    return (validator.overrideSymbol(NAME) as string).toUpperCase()
  }

  return async () => {
    const defaultConfig = config || makeConfig()

    const ws = new IntrinioRealtime({
      api_key: defaultConfig.apiKey,
      provider: 'iex',
    })

    await ws._refreshToken()
    ws.destroy() // close WS connection generated by the Intrinio SDK to prevent unmaintained connections

    return {
      connection: {
        url: ws._makeSocketUrl(),
      },
      subscribe: (input) => ws._makeJoinMessage(getBase(input)),
      unsubscribe: (input) => ws._makeLeaveMessage(getBase(input)),
      subsFromMessage: (message) => ws._makeJoinMessage(message.payload.ticker),
      isError: (message: any) => Number(message.TYPE) > 400 && Number(message.TYPE) < 900,
      filter: (message) => message.event == 'quote' && message.payload?.type == 'last',
      toResponse: (wsResponse: any): AdapterResponse =>
        Requester.success(undefined, { data: { result: wsResponse?.payload?.price } }),
    }
  }
}
