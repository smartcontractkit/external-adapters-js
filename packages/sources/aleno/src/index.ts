import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { config } from './config'
import { price } from './endpoint'

export const adapter = new Adapter({
  defaultEndpoint: price.name,
  name: 'ALENO',
  config,
  endpoints: [price],
  rateLimiting: {
    tiers: {
      default: {
        rateLimit1m: 30,
      },
    },
  },
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
