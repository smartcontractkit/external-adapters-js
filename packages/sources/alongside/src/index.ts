import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { collateral } from './endpoint'
import { config } from './config'

export const adapter = new Adapter({
  defaultEndpoint: collateral.name,
  name: 'ALONGSIDE',
  config,
  endpoints: [collateral],
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
