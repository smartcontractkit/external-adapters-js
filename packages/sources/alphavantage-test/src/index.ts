import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { PriceAdapter } from '@chainlink/external-adapter-framework/adapter'
import { config } from './config'
import { forex } from './endpoint'

export const adapter = new PriceAdapter({
  defaultEndpoint: forex.name,
  name: 'ALPHAVANTAGE',
  config,
  endpoints: [forex],
  rateLimiting: {
    tiers: {
      free: {
        rateLimit1m: 5,
        rateLimit1h: 20.83,
      },
      '49.99': {
        rateLimit1m: 75,
      },
      '99.99': {
        rateLimit1m: 150,
      },
      '149.99': {
        rateLimit1m: 300,
      },
      '199.99': {
        rateLimit1m: 600,
      },
      '249.99': {
        rateLimit1m: 1200,
      },
    },
  },
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
