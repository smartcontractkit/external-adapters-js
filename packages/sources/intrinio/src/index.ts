import { expose } from '@chainlink/ea-bootstrap'
import { makeExecute, makeWSHandler } from './adapter'
import { makeConfig, NAME } from './config'

export = { NAME, makeExecute, makeConfig, ...expose(makeExecute(), makeWSHandler()) }
