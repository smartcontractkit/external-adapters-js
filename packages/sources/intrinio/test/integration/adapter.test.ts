import { AdapterRequest, FastifyInstance, util } from '@chainlink/ea-bootstrap'
import request, { SuperTest, Test } from 'supertest'
import * as process from 'process'
import { server as startServer } from '../../src'
import * as nock from 'nock'
import {
  mockAuthResponse,
  mockRateResponseSuccess,
  mockSubscribeResponse,
  mockUnsubscribeResponse,
} from './fixtures'
import { AddressInfo } from 'net'
import {
  mockWebSocketProvider,
  mockWebSocketServer,
  MockWsServer,
  mockWebSocketFlow,
  setEnvVariables,
  setupExternalAdapterTest,
} from '@chainlink/ea-test-helpers'
import type { SuiteContext, TestOptions } from '@chainlink/ea-test-helpers'
import { WebSocketClassProvider } from '@chainlink/ea-bootstrap/dist/lib/middleware/ws/recorder'

describe('execute', () => {
  const id = '1'
  const context: SuiteContext = {
    req: null,
    server: startServer,
  }

  const options: TestOptions = {
    cleanNock: false,
    fastify: false,
  }
  const envVariables = {
    API_KEY: process.env.API_KEY || 'fake-api-key',
  }

  setupExternalAdapterTest(envVariables, context, options)

  describe('price api', () => {
    const data: AdapterRequest = {
      id,
      data: {
        base: 'AAPL',
      },
    }

    it('should return success', async () => {
      mockRateResponseSuccess()

      const response = await (context.req as SuperTest<Test>)
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })
  })
})

describe('websocket', () => {
  let mockedWsServer: InstanceType<typeof MockWsServer>
  let fastify: FastifyInstance
  let req: SuperTest<Test>

  let oldEnv: NodeJS.ProcessEnv
  beforeAll(async () => {
    if (!process.env.RECORD) {
      process.env.API_KEY = 'fake-api-key'
      mockedWsServer = mockWebSocketServer(
        `wss://realtime.intrinio.com/socket/websocket?vsn=1.0.0&token=fake-api-token`,
      )
      mockWebSocketProvider(WebSocketClassProvider)
    }

    oldEnv = JSON.parse(JSON.stringify(process.env))
    process.env.WS_ENABLED = 'true'
    process.env.WS_SUBSCRIPTION_TTL = '300'

    fastify = await startServer()
    req = request(`localhost:${(fastify.server.address() as AddressInfo).port}`)
  })

  afterAll((done) => {
    setEnvVariables(oldEnv)
    nock.restore()
    nock.cleanAll()
    nock.enableNetConnect()
    fastify.close(done)
  })

  describe('iex endpoint', () => {
    const jobID = '1'

    it('should return success', async () => {
      const data: AdapterRequest = {
        id: jobID,
        data: {
          endpoint: 'iex',
          base: 'AAPL',
        },
      }

      let flowFulfilled = Promise.resolve(true)
      if (!process.env.RECORD) {
        mockAuthResponse()
        mockRateResponseSuccess()

        flowFulfilled = mockWebSocketFlow(mockedWsServer, [
          mockSubscribeResponse,
          mockUnsubscribeResponse,
        ])
      }

      const makeRequest = () =>
        req
          .post('/')
          .send(data)
          .set('Accept', '*/*')
          .set('Content-Type', 'application/json')
          .expect('Content-Type', /json/)
          .expect(200)

      // We don't care about the first response, coming from http request
      // This first request will start both batch warmer & websocket
      await makeRequest()

      // This final request should disable the cache warmer, sleep is used to make sure that the data is  pulled from the websocket
      // populated cache entries.
      await util.sleep(100)
      const response = await makeRequest()

      expect(response.body).toEqual({
        jobRunID: '1',
        result: 166.91,
        statusCode: 200,
        maxAge: 30000,
        data: { result: 166.91 },
      })

      await flowFulfilled
    }, 30000)
  })
})
