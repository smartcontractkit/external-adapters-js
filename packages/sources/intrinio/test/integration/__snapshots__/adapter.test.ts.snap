// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute price api should return success 1`] = `
{
  "data": {
    "result": 24.25,
  },
  "jobRunID": "1",
  "providerStatusCode": 200,
  "result": 24.25,
  "statusCode": 200,
}
`;
